### PR TITLE
[SPARK-7232] [SQL] Add a Substitution batch for spark sql analyzer

### PR DESCRIPTION
**Pull request by scwf**
Added a new batch named `Substitution` before `Resolution` batch. The motivation for this is there are kind of cases we want to do some substitution on the parsed logical plan before resolve it. 
Consider this two cases:
1 CTE, for cte we first build a row logical plan

```
'With Map(q1 -> 'Subquery q1
                   'Project ['key]
                      'UnresolvedRelation [src], None)
 'Project [*]
  'Filter ('key = 5)
   'UnresolvedRelation [q1], None
```

In `With` logicalplan here is a map stored the (`q1-> subquery`), we want first take off the with command and substitute the  `q1` of `UnresolvedRelation` by the `subquery`

2 Another example is Window function, in window function user may define some windows, we also need substitute the window name of child by the concrete window. this should also done in the Substitution batch.

### CODE REVIEW

**Comment by scwf** [comment]
Retest this please

**Comment by marmbrus** [comment]
@yhuai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -69,6 +72,38 @@ class Analyzer(
   )
 
   /**
+   * Substitute CTE definitions
+   */
+  object CTESubstitution extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = {
+      val (realPlan, cteRelations) = plan match {
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Do we still need to first get `realPlan` and `creRelations`? Can we just use the following? 
```
def apply(plan: LogicalPlan): LogicalPlan = plan match {
  case With(child, relations) => substituteCTE(child, relations)
  case other => other 
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -69,6 +72,38 @@ class Analyzer(
...
+    def substituteCTE(plan: LogicalPlan, cteRelations: Map[String, LogicalPlan]): LogicalPlan = {
+      plan transform {
+        case i @ InsertIntoTable(u: UnresolvedRelation, _, _, _, _) =>
+          // In hive, if there is same table name in database and CTE definition,
+          // hive will use the table in database, not the CTE one.
+          // Taking into account the reasonableness and the implementation complexity,
+          // here use the CTE definition first, check table name only and ignore database name
+          val relation = cteRelations.get(u.tableIdentifier.last)
+            .map(relation => u.alias.map(Subquery(_, relation)).getOrElse(relation))
+            .getOrElse(u)
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Maybe the following will be easier to read?
```
val substituted = cteRelations.get(u.tableIdentifier.last).map { relation =>
  val withAlias = u.alias.map(Subquery(_, relation)
  withAlias.getOrElse(relation)
}
substituted.getOrElse(u)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -46,13 +46,16 @@ class Analyzer(
   val resolver = if (caseSensitive) caseSensitiveResolution else caseInsensitiveResolution
 
   val fixedPoint = FixedPoint(maxIterations)
-
+
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Unnecessary change.

**Comment by yhuai** [comment]
Thanks for working on it. I have left a few minor comments.

**Comment by scwf** [comment]
@yhuai ,thanks, updated!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -69,6 +72,36 @@ class Analyzer(
...
+          // In hive, if there is same table name in database and CTE definition,
+          // hive will use the table in database, not the CTE one.
+          // Taking into account the reasonableness and the implementation complexity,
+          // here use the CTE definition first, check table name only and ignore database name
+          val substituted = cteRelations.get(u.tableIdentifier.last).map { relation =>
+            val withAlias = u.alias.map(Subquery(_, relation))
+            withAlias.getOrElse(relation)
+          }
+          substituted.getOrElse(u)
+          i.copy(table = relation)
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: ` i.copy(table = substituted.getOrElse(u))`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -69,6 +72,36 @@ class Analyzer(
...
+        case u : UnresolvedRelation =>
+          cteRelations.get(u.tableIdentifier.last)
+            .map(relation => u.alias.map(Subquery(_, relation)).getOrElse(relation))
+            .getOrElse(u)
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: You may want to also update this part.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 4 above)
Comment (scwf) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: oh, my bad, thanks

**Comment by scwf** [comment]
Retest this please.

**Comment by scwf** [comment]
Updated.

**Comment by scwf** [comment]
@yhuai  any more comments?

**Comment by yhuai** [comment]
LGTM

**Comment by scwf** [comment]
@marmbrus can you merge this? #5604 need refactory based on this.

**Comment by scwf** [comment]
ping @marmbrus

**Comment by scwf** [comment]
Retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -69,6 +72,37 @@ class Analyzer(
...
+   */
+  object CTESubstitution extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan match {
+      case With(child, relations) => substituteCTE(child, relations)
+      case other => other
+    }
+
+    def substituteCTE(plan: LogicalPlan, cteRelations: Map[String, LogicalPlan]): LogicalPlan = {
+      plan transform {
+        case i @ InsertIntoTable(u: UnresolvedRelation, _, _, _, _) =>
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I feel this rule does nothing.... We are inserting to a table and this table is actually a CTE?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -169,36 +203,20 @@ class Analyzer(
...
       } catch {
         case _: NoSuchTableException =>
           u.failAnalysis(s"no such table ${u.tableName}")
       }
     }
 
-    def apply(plan: LogicalPlan): LogicalPlan = {
-      val (realPlan, cteRelations) = plan match {
-        // TODO allow subquery to define CTE
-        // Add cte table to a temp relation map,drop `with` plan and keep its child
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Should we keep the TODO at here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 6 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I feel we should get rid of this rule. I do not think this one actually works.

**Comment by yhuai** [comment]
@scwf Can you also move the the rule for handling window clause (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L668-680) to this batch?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 6 above)
Comment (scwf) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: yes, i will delete this case

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 7 above)
Comment (scwf) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: yes, we should keep it

**Comment by scwf** [comment]
Updated done /cc @yhuai

**Comment by yhuai** [comment]
Thanks! Merging to master and branch 1.4.
